src/libcore/iter/iterator.rs

```diff
@@ -1696,6 +1736,41 @@ pub trait Iterator {
         self.try_fold(init, move |acc, x| Ok::<B, !>(f(acc, x))).unwrap()
     }
 
+    /// The same as [`fold()`](#method.fold), but uses the first element in the
+    /// iterator as the initial value, folding every subsequent element into it.
+    /// If the iterator is empty, return `None`; otherwise, return the result
+    /// of the fold.
+    ///
+    /// # Example
+    ///
+    /// Find the maximum value:
+    ///
+    /// ```
+    /// fn find_max<I>(iter: I) -> Option<I::Item>
+    ///     where I: Iterator,
+    ///           I::Item: Ord,
+    /// {
+    ///     iter.fold_self(|a, b| {
+    ///         a.partial_cmp(b).map(move |cmp| match cmp {
+    ///             Ordering::Greater | Ordering::Equal => a,
+    ///             Ordering::Less => b,
+    ///         })
+    ///     })
+    /// }
+    /// let a = [10, 20, 5, -23, 0];
+    /// let b = [];
+    ///
+    /// assert_eq!(find_max(a.iter()), Some(20));
+    /// assert_eq!(find_max(b.iter()), None));
+    /// ```
+    #[inline]
+    #[unstable(feature = "iterator_fold_self", issue = "60103")]
+    fn fold_self<F>(mut self, mut f: F) -> Option<Self::Item>
```

Comment:
For reference, in itertools, this is https://docs.rs/itertools/0.8.0/itertools/trait.Itertools.html#method.fold1 which is inspired by https://hackage.haskell.org/package/base-4.12.0.0/docs/Prelude.html#v:foldl1 in Haskell.